### DISCUSSION TITLE
You can avoid errors that can occur as a result of updating data (in the database) that are in the self.scope that the code may depend on as these data won't be updated

### DISCUSSION THREAD

**Discussion by ahmedyasserays** [django/channels#1742] [Help!] [answered]
I'm building an app where users can play online together , after each game ends I should save the win and lose counter for each player . 
let's consider two players a and b playing together 
at first game player a won and his score is now 1 and player b lose score is now 1
second game player a also won now his score is 2 and player b lose score is 2
at the third game player b won so his score should be 1 win and 2 loses , for player a he should be 2 wins and 1 lose
this doesn't happen and the player a win score returns back to zero , the lose counter of player b won't be reset 

I'm writing sync code and I don't have any idea why is this happening 

Note : i'm using this weird competitor thing to prevent any one from sending that he won to the socket , so only the user can confirm the competitors has won

my consumers.py
```python

    def receive(self, text_data):
        player = Room.objects.get(channel_name=self.room_group_name).presence_set.filter(channel_name=self.channel_name)
        if not player.exists():
            self.send(json.dumps({
                'type':'error',
                'code':'408',
                'error':'Time out! You have to play in 15 seconds'
                }))
            self.close()
            return

        text_data = json.loads(text_data)
        if text_data['type'] ==  'completed':
            competitor:CustomUser = Room.objects.get(channel_name=self.room_group_name).presence_set.filter(~Q(channel_name=self.channel_name)).first().user
            user:CustomUser = self.scope['user']
            db_logger.info(f'Game ended between me {user} and {competitor}\nuser scores: {user.won_games}, {user.lost_games}, {user.draw_games}\ncompetitor scores: {competitor.won_games}, {competitor.lost_games}, {competitor.draw_games}')
            if competitor:
                competitor.win()
            if user.is_authenticated:
                user.lose()
            db_logger.info(f'Game saved between me {user} and {competitor}\nuser scores: {user.won_games}, {user.lost_games}, {user.draw_games}\ncompetitor scores: {competitor.won_games}, {competitor.lost_games}, {competitor.draw_games}')
            return 
        elif text_data['type'] == 'draw':
            competitor:CustomUser = Room.objects.get(channel_name=self.room_group_name).presence_set.filter(~Q(channel_name=self.channel_name)).first().user
            if competitor:
                competitor.draw_games += 1
                competitor.save()
            user = self.scope['user']
            if user.is_authenticated:
                user.draw_games += 1
                user.save()           
            return
# some code to handle some other types
```

models.py
```python

class CustomUser(AbstractUser):
    won_games = models.IntegerField(default=0)
    lost_games = models.IntegerField(default=0)
    draw_games = models.IntegerField(default=0)




    def win(self):
        self.won_games += 1
        self.save() 

    def lose(self):
        self.lost_games += 1
        self.save() 

    def draw(self):
        self.draw_games += 1
        self.save() 
```

**Comment by ahmedyasserays** [comment, marked as answer]
so the problem was in this line
```python
user:CustomUser = self.scope['user']
```
which uses the scope of the instance of the consumer , this scope has just a copy of information about the user at the initiation point which means updating any info about the user after the connection wouldn't affect the user in the scope which mean saving this user object will override the new updated data which exactly what happened and could be fixed by re fetching the user object from the database so the code should be 
```python
user:CustomUser = self.scope['user']
user = CustomUser.objects.get(pk=user.pk)
```
and every thing is fine now .
I haven't read the source code , but I suggest that you can change the self.scope attribute to be a property method which fetch the user data from the database or any data that could be updated after the connection is made and the code may depend on it to avoid errors like this .

**Reply by Amohammadi2** [reply]
you can also use `user.refresh_from_db()` I think.

**Reply by ahmedyasserays** [reply]
Yes you're right , and we can use the F() class to update only single field in the database to avoide race conditions like this as mentioned [here](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#avoiding-race-conditions-using-f)
@Amohammadi2
